include/parameter.h

```diff
@@ -0,0 +1,307 @@
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef PARAMETERS_PARAMETER_H
+#define PARAMETERS_PARAMETER_H
+
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <memory>
+#include <set>
+#include <stdexcept>
+#include <string>
+
+#ifndef PARAMETERS_LIMITS_H
+#include "limits.h"
+#endif
+
+#ifndef PARAMETERS_TRANSFORM_H
+#include "transform.h"
+#endif
+
+#ifndef PARAMETERS_TYPE_NAME_H
+#include "type_name.h"
+#endif
+
+namespace parameters {
+
+class Flag {
+private:
+    static inline const std::string _name = "";
+    bool _value;
+
+public:
+    static const std::string get_name() { return _name; }
+    bool get_value() const { return _value; }
+
+    void set_value(bool value) { _value = value; }
+
+    Flag(bool value) : _value(value) {}
+};
+
+template<typename T, class Unit>
+class ParameterBase {
+public:
+    virtual bool get_fixed() const;
+    virtual bool get_free() const;
+    virtual std::string get_label() const;
+    virtual inline const Limits<T> & get_limits() const;
+    virtual inline const Transform<T> & get_transform() const;
+    virtual inline T get_transform_derivative() const;
+    virtual inline const std::shared_ptr<const Unit> get_unit() const;
+    virtual T get_value() const;
+    virtual T get_value_transformed() const;
+
+    virtual void set_fixed(bool fixed);
+    virtual void set_free(bool free);
+    virtual void set_label(std::string label);
+    virtual void set_limits(std::shared_ptr<const Limits<T>> limits);
+    virtual void set_transform(const std::shared_ptr<const Transform<T>> transform);
+    virtual void set_value(T value);
+    virtual void set_value_transformed(T value_transformed);
+    virtual void set_unit(std::shared_ptr<const Unit> unit = nullptr);
+    virtual std::string str() const;
+
+
+    virtual ~ParameterBase() = default;
+};
+
+template<typename T, class C, class Unit>
+class Parameter : public ParameterBase<T, Unit>, public std::enable_shared_from_this<C> {
+public:
+    using SetC = std::set<std::shared_ptr<C>>;
+
+private:
+    // TODO: Remember why this class needs to exist, if it does
+    // Probably just to store the const Limits reference?
```

Comment:
I think it exists so that I can set the transform to a static const(expr) like limits_maximal. If there's a better way to accomplish something like that (possibly with enable_shared_from_this?), do let me know.